### PR TITLE
chore: production environment setup

### PR DESCRIPTION
**Pull request by DorisWongNHS**
add production deployment workflow and task definition for production

TISNEW-5487

### CODE REVIEW

**Review by uzairbaig — DISMISSED** [review]

File: .github/workflows/prod.deploy.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        id: login-ecr
+        uses: aws-actions/amazon-ecr-login@v1
+
+      - name: Fill in the new image ID in the Amazon ECS task definition
+        id: task-def
+        uses: aws-actions/amazon-ecs-render-task-definition@v1
+        with:
+          task-definition: .aws/task-definition-prod.json
+          container-name: ${{ github.event.repository.name }}
+          image: ${{ steps.login-ecr.outputs.registry }}/${{ github.event.repository.name }}:${{ github.sha }}
```
Comment (Judge40) [inline .github/workflows/prod.deploy.yml:31]: No guarantee that the commit passed stage pipeline/tests

File: .github/workflows/prod.deploy.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+      - name: Deploy Amazon ECS task definition
+        uses: aws-actions/amazon-ecs-deploy-task-definition@v1
+        with:
+          task-definition: ${{ steps.task-def.outputs.task-definition }}
+          service: ${{ github.event.repository.name }}
+          cluster: revalidation-prod
+          wait-for-service-stability: false
```
Comment (Judge40) [inline .github/workflows/prod.deploy.yml:39]: Needs to be set to `true`

Comment (DorisWongNHS) [reply .github/workflows/prod.deploy.yml]: updated.

File: .github/workflows/prod.deploy.yml (hunk 1 above)
Comment (DorisWongNHS) [reply .github/workflows/prod.deploy.yml]: One more step is added to the stage pipeline for adding a tag as 'stable' after deployment, and also changed the image we upload to Prod as 'stable'

**Review by uzairbaig — APPROVED** [review]

**Review by uzairbaig — APPROVED** [review]

File: .aws/task-definition-prod.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+  ],
+  "executionRoleArn": "ecsTaskExecutionRole",
+  "family": "tis-revalidation-concerns-prod",
+  "requiresCompatibilities": [
+    "FARGATE"
+  ],
+  "networkMode": "awsvpc",
+  "cpu": "256",
+  "memory": "4096"
```
Comment (Judge40) [inline .aws/task-definition-prod.json:64]: Should be 1G

File: .github/workflows/ci-cd-workflow.yml [hunk 4]
```diff
@@ -191,7 +191,16 @@ jobs:
...
           cluster: revalidation-preprod
-          wait-for-service-stability: false
+          wait-for-service-stability: true
+
+      - name: Build, tag as stable and push image to Amazon ECR
+        env:
+          ECR_REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+          ECR_REPOSITORY: ${{ github.event.repository.name }}
+        run: |
+          ./gradlew bootBuildImage -x bootJar --imageName=$ECR_REGISTRY/$ECR_REPOSITORY:${{ github.sha }}
```
Comment (Judge40) [inline .github/workflows/ci-cd-workflow.yml:201]: Don't rebuild the image, tag the existing one.

**Review by Judge40 — CHANGES_REQUESTED** [review]

File: .aws/task-definition-prod.json (hunk 3 above)
Comment (DorisWongNHS) [reply .aws/task-definition-prod.json:64]: Do we need to match it with the task definition in TIS-OPS and the ECS main.tf? They are both 4096 for memory.

https://github.com/Health-Education-England/TIS-OPS/blob/master/terraform/environments/prod/revalidation/ecs/task-definitions/concerns.json

https://github.com/Health-Education-England/TIS-OPS/blob/master/terraform/environments/prod/revalidation/ecs/main.tf

Comment (Judge40) [reply .aws/task-definition-prod.json:64]: Everything should match what is in this project's task-definition file.

Comment (DorisWongNHS) [reply .aws/task-definition-prod.json]: Updated to 1024. Thanks.

**Review by Judge40 — APPROVED** [review]

**Pull request merged**